### ISSUE TITLE
Keine Negativlagerung möglich obwohl in Mandanteneinstellungen aktiviert

### ISSUE THREAD

**Issue by davidhommel** [selectline-software/selectline-api#272]
Wenn ich über die API eine Artikelposition mit einem Artikel ohne Lagerbestand in einen Lieferschein schreibe, bekomme ich eine Fehlermeldung bzgl. einer negativen offenen Menge. Wenn ich für den betroffenen Artikel eine manuelle Einlagerung mache, sodass Bestand vorhanden ist, funktioniert das Anlegen der Position.

Allerdings ist in den Mandanteneinstellungen die Option ,,Negativ lagern zulassen'' aktiv - Die API wirft aber dennoch den Fehler. 

Hier die Fehlermeldung:
![grafik](https://user-images.githubusercontent.com/72391062/200837600-a695f60b-4bca-4140-aa2a-f79eec8fa812.png)

**Comment by ThePholph** [comment]
Hallo,

kannst du den Request zur Verfügung stellen?

Vielen Grüße

Daniel

**Comment by davidhommel** [comment]
Klaro:
`2022-11-10 13:55:30	|VERB	|SL_PutBelege	|CP-	|API Aufruf zur Erstellung Belegposition: 
POST /slmobileApi/Documents/LLS823/ArticleItem
{
"DeliveryDate": "2022-11-10T13:53:27.0000000+01:00",
"CalculatedQuantityValue": "1",
"ArticleNumber": "10079",
"Name": "LevelOne Netzteil 0R10028  12V / 1,50A DC Power Ersatzteil",
"AdditionalDescription": "LevelOne Netzteil 0R10028  12V / 1,50A DC Power Ersatzteil",
"Memo": "besoMobile-Beleg: 184400001"
}
`

**Comment by davidhommel** [comment]
Der Vollständigkeit halber auch noch die auf diesen Request zurückgemeldete Fehlernachricht
```
2022-11-10 13:55:31	|ERR	|SL_PutBelege	|CP12	|Position 18440000100005 nicht in SelectLine übertragen: Server answered with an error!
500: Internal Server Error{
  "StatusCode": "InternalServerError",
  "ResponseCode": "42-000",
  "ResponseId": 0,
  "ResponseIdDescription": "Undefined",
  "Message": "Fehler beim Lagern. Offene Menge: -1,00",
  "Details": {
    "SQLFehlerNr": "0",
    "ExtMessage": "",
    "SQLTmpID": "0",
    "ehlerNr": "102065",
    "ExceptionName": "ELagerungException"
  }
}
```

**Comment by MatthiasGuse** [comment]
Hallo,

wurde die Option für die negative Lagerung erst gesetzt? Bzw. wurde nach einer Änderung an der Option der API User vollständig ab- und wieder angemeldet? Alternativ der Backendmanager neu gestartet?

Das Verhalten klingt so, dass die Option im Bakend noch auf einem alten Stand im Speicher ist.

Viele Grüße

**Comment by davidhommel** [comment]
Guten Morgen,

Die Einstellung war von Anfang an im Mandanten gesetzt. Habe zur Kontrolle aber nochmal das API-Backend neu gestartet - gleicher Fehler...

**Comment by MatthiasGuse** [comment]
Ein paar Fragen:

1. Welche Version ist im Einsatz?
2. Hat das überhaupt schon mal in diesem Mandanten funktioniert?
3. Steht das Auftreten im Zusammenhang mit einem Update?
4. Hat der Artikel zusätzliche Eigenschaften wie Serie / Charge oder MHD?

**Comment by ThePholph** [comment]
Hallo,

ich kann den Fehler reproduzieren, mit "alten" Artikeln, bei denen der Hacken "Lagerdialog immer anzeigen" gesetzt und ausgegraut ist. Bei neu angelegten Artikeln, kann ich es reproduzieren, wenn bei "Lagerstrategie" "keine" gewählt wird (hier ist es egal, ob "Lagerdialog immer anzeigen" gesetzt ist oder nicht). Ohne wesentlich mehr Aufwand ist nicht ganz 100 % auszumachen, woran es genau liegt. 

Ich bin bisher nicht auf den Fehler gestoßen, da wir aktuell via API keine Positionen in Lieferscheine einfügen ohne Bezug auf einen Auftrag.

Ich hoffe, ich konnte etwas weiterhelfen.

Viele Grüße

Daniel

**Comment by MatthiasGuse** [comment]
Danke Daniel,

du hast den richtigen Anstoß gegeben.

Problem sind die fehlenden Lagerinformationen. Da es sich um einen lagernden Beleg handelt und auch bei dem Artikel in der Warenwirtschaft ein Lagerdialog aufgehen würde, kann die API den Artikel auch nicht lagern. Es kann kein Lager ermittelt werden. 
Daher sind die `StoreInformation` an dieser Stelle zwingend zu füllen, um das gewünschte Lager, in dem der negative Bestand entstehen soll, mitzugeben.

Siehe Swagger Beispiel:

 {
  "ArticleNumber": "100001",
  "UnitPrice": 89.99,
  "AdditionalDescription": "Schaukelstuhl, blau, breite Armlehnen",
  "CalculatedQuantityValue": 4,
  "StoreInformation": [
    {
      "ArticleNumber": "100001",
      "Warehouse": "200",
      "Quantity": 2
    },
    {
      "ArticleNumber": "100001",
      "Warehouse": "400",
      "Quantity": 2
    }
  ]
}

Grüße

**Comment by ThePholph** [comment]
Hallo Matthias,

das klingt logisch und nachvollziehbar.

Idee hierzu:
Vereinfachen würde es sich für uns, wenn hier das Standardlager greifen würde (natürlich nur, wenn es hinterlegt ist).

Grüße 

Daniel

**Comment by MatthiasGuse** [comment]
Ja, ich würde behaupten, dass es genauso funktioniert und dass das bei @davidhommel nicht der Fall ist. Die Einstellungen der Warenwirtschaft sind so gesetzt, dass in der vorhandenen Konstellation kein Lager ermittelt wird und deshalb der Lagerdialog aufgeht. Geht der Lagerdialog nicht auf, funktioniert es sicher auch per API. Geht er auf, streikt auch die API.

Daher ist es immer besser die Lagerinformationen gleich mitzugeben, wenn man wirklich in einem lagernden Beleg unterwegs ist. Der Lagerdialog ist für die Warenwirtschaft sozusagen die Rückfalloption, wenn nichts ermittelt oder vorgegeben wird. Da die API keinen Lagerdialog hat, endet der Weg in der gezeigten Fehlermeldung.

Grüße

**Comment by davidhommel** [comment]
Guten Morgen,
Der Fehler verschwindet tatsächlich beim Deaktivieren der Option, dass der Lagerdialog beim Auslagern immer angezeigt werden soll - und wenn alle Artikel ein Standardlager hinterlegt haben. Das habe ich gestern mit einem SQL Skript für alle Artikel nachgepflegt und siehe da: Es funktioniert.

Danke @ThePholph und @MatthiasGuse für die Unterstützung!
